buckets/local/add.go

```diff
@@ -0,0 +1,213 @@
+			return err
+		}
+	}
+
+	// Add files that are missing, or were decided to be overwritten.
+	if len(toAdd) > 0 {
+		if events != nil {
+			events <- PathEvent{
+				Path: ipfsBasePth.String(),
+				Type: PathStart,
+			}
+		}
+		eg, gctx := errgroup.WithContext(ctx)
+		for _, o := range toAdd {
+			o := o
+			eg.Go(func() error {
+				if gctx.Err() != nil {
+					return nil
+				}
+				if err := os.Remove(o.path); err != nil && !os.IsNotExist(err) {
+					return err
+				}
+				trimmedDest := strings.TrimPrefix(o.path, dest)
+				return b.getIpfsFile(gctx, path.Join(ipfsBasePth, trimmedDest), o.path, o.size, o.cid, events)
+			})
+		}
+		if err := eg.Wait(); err != nil {
+			return err
+		}
+		if events != nil {
+			events <- PathEvent{
+				Path: ipfsBasePth.String(),
+				Type: PathComplete,
+			}
+		}
+	}
+	return nil
+}
+
+// listMergePath walks the local bucket and the remote IPFS UnixFS DAG asking
+// the client if wants to (replace, merge, ignore) matching folders, and if wants
+// to (overwrite, ignore) matching files. Any non-matching files or folders in the
+// IPFS UnixFS DAG will be added locally.
+// The first return value is a slice of paths of folders that were decided to be
+// replaced completely (not merged). The second return value are a list of files
+// that should be added locally. If one of them exist, can be understood that should
+// be overwritten.
+func (b *Bucket) listMergePath(ctx context.Context, ipfsBasePth path.Path, ipfsRelPath, dest string, merge SelectMergeFunc) ([]string, []object, error) {
+	// List remote IPFS UnixFS path level
+	rep, err := b.clients.Buckets.ListIpfsPath(ctx, path.Join(ipfsBasePth, ipfsRelPath))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// If its a dir, ask if should be ignored, replaced, or merged.
+	if rep.Item.IsDir {
+		var replacedFolders []string
+		var toAdd []object
+
+		var folderExists bool
+
+		localFolderPath := filepath.Join(dest, ipfsRelPath)
+		if _, err := os.Stat(localFolderPath); err == nil {
+			folderExists = true
+		}
+
+		if folderExists && merge != nil {
+			ms, err := merge(fmt.Sprintf("Merge strategy for  %s", localFolderPath), true)
+			if err != nil {
+				return nil, nil, err
+			}
+			switch ms {
+			case Skip:
+				return nil, nil, nil
+			case Merge:
+				break
+			case Replace:
+				replacedFolders = append(replacedFolders, localFolderPath)
+				merge = nil
+			}
+		}
+		for _, i := range rep.Item.Items {
+			nestFolderReplace, nestAdd, err := b.listMergePath(ctx, ipfsBasePth, filepath.Join(ipfsRelPath, i.Name), dest, merge)
+			if err != nil {
+				return nil, nil, err
+			}
+			replacedFolders = append(replacedFolders, nestFolderReplace...)
+			toAdd = append(toAdd, nestAdd...)
+		}
+		return replacedFolders, toAdd, nil
+	}
+
+	// If it's a file and it exists, confirm whether or not it should be overwritten.
+	pth := filepath.Join(dest, ipfsRelPath)
+	if _, err := os.Stat(pth); err == nil && merge != nil {
```

Comment:
Instead of dealing with the cmd prompt, this uses a user-supplied method for selecting the merge strategy.